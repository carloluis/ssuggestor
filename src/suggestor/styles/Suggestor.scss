:global .ss-triangle,
:global .ss-remove {
	position: absolute;
	cursor: pointer;
	margin: auto;
	color: #ccc;
	font-size: 16px;
	right: 10px;
	top: 0;
	bottom: 0;
	height: 14px;
	z-index: 4;
}

:global .ss-remove {
	font-size: 14px;
	right: 30px;
}
