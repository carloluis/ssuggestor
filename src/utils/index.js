import noop from './noop';
import removeAccents from './remove-accents';
import withClickOut from './withClickOut';

export { EMPTY_STR, KEY_CODES } from './values';
export {
	noop,
	removeAccents,
	withClickOut
};
