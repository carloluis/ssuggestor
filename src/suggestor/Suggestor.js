import React, { PureComponent } from 'react';
import { EMPTY_STR, KEY_CODES } from '../utils/values';
import { SPIN_STYLES, X_STYLES } from './styles';
import withClickOut from '../utils/withClickOut';
import List from './List';

export class Suggestor extends PureComponent {
	constructor(props){
		super(props);
		this._bind('handleClick', 'handleChange', 'remove', 'handleKeyDown', 'handleItemClick', 'handleItemMouseEnter', 'handleClickOut', 'focus');

		this.state = {
			value: props.value,
			filtered: props.list,
			open: false,
			index: 0
		};
	}
	_bind(...methods) {
		methods.forEach(method => this[method] = this[method].bind(this));
	}
	handleClickOut() {
		this.handleClose();
	}
	componentWillReceiveProps(nextProps) {
		if (nextProps.value !== this.state.value) {
			this.setState({ value: nextProps.value });
		}
	}
	handleClose() {
		this.setState({ open: false, filtered: this.props.list, index: 0 });
	}
	toggleList() {
		if (this.state.open) {
			this.handleClose();
		}
		else {
			this.setState({
				open: true
			});
		}
	}
	handleClick() {
		if (this.props.openOnClick) {
			this.toggleList();
		}
	}
	handleKeyDown(e) {
		this.props.onKey(e);
		if (!this.props.useKeys) {
			return;
		}
		let { open, index, filtered:list, value } = this.state;

		switch (e.keyCode) {
			case KEY_CODES.TAB:
				if (this.props.selectOnTab && open && list[index]) {
					this.changeValue(list[index], true);
				}
				this.handleClose();
				return;
			case KEY_CODES.ENTER:
				this.toggleList();
				if (open && list[index]) {
					this.changeValue(list[index], true);
				} else {
					this.props.onSelect(value);
				}
				break;
			case KEY_CODES.ESCAPE:
				this.handleClose();
				if (!open) {
					this.changeValue(EMPTY_STR);
				}
				break;
			case KEY_CODES.DOWN: {
				let next = (index + open) % list.length;
				this.setState({ open:true, index: next });
				break;
			}
			case KEY_CODES.UP: {
				let prev = (index || list.length) - 1;
				this.setState({ open: true, index: prev });
				break;
			}
			default:
				return;
		}

		e.preventDefault();
	}
	handleItemClick(value) {
		this.changeValue(value, true);
		if(!this.props.openOnClick) {
			this.toggleList();
		}
	}
	handleItemMouseEnter(index) {
		this.setState({ index });
	}
	handleChange(e) {
		e.stopPropagation();
		let value = e.target.value;
		let suggest = value.length >= this.props.suggestOn;
		let filtered = this.filter(value);
		let nextState = { filtered };
		if(!filtered.length || !suggest) {
			this.handleClose();
		} else{
			nextState.open = true;
		}
		this.setState(nextState);
		this.changeValue(value);
	}
	remove() {
		this.changeValue(EMPTY_STR, true);
	}
	changeValue(value, select=false) {
		this.setState({ value }, () => {
			this.props.onChange(value);
			if(select) {
				this.props.onSelect(value);
			}
		});
	}
	filter(value) {
		value = value.toLowerCase();
		return this.props.list.filter(item => item.toLowerCase().indexOf(value) !== -1);
	}
	focus() {
		this.refs.input.focus();
	}
	render() {
		let { className, style, placeholder, arrow, close, tooltip, required } = this.props;
		let { open, value, index, filtered:list } = this.state;

		return (
			<div className={className} style={style} onClick={this.handleClick} onKeyDown={this.handleKeyDown} ref={this.props.reference} >
				<input type="text" className="form-control" onChange={this.handleChange} value={value} ref="input" placeholder={placeholder} title={tooltip} required={required} />
				{ arrow && <span className="glyphicon glyphicon-triangle-bottom" style={SPIN_STYLES} /> }
				{ close && value && <span className="glyphicon glyphicon-remove" style={X_STYLES} onClick={this.remove}/> }
				<List {...{ list, index, open, value }} onItemClick={this.handleItemClick} onItemMouseEnter={this.handleItemMouseEnter} />
			</div>
		);
	}
}

Suggestor.propTypes = {
	list: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
	reference: React.PropTypes.func,
	onChange: React.PropTypes.func,
	onSelect: React.PropTypes.func,
	onKey: React.PropTypes.func,
	value: React.PropTypes.string,
	openOnClick: React.PropTypes.bool,
	selectOnTab: React.PropTypes.bool,
	placeholder: React.PropTypes.string,
	tooltip: React.PropTypes.string,
	className: React.PropTypes.string,
	suggestOn: React.PropTypes.number,
	style: React.PropTypes.object,
	required: React.PropTypes.bool,
	useKeys: React.PropTypes.bool,
	arrow: React.PropTypes.bool,
	close: React.PropTypes.bool
};

const nop = _ => _;

Suggestor.defaultProps = {
	className: 'input-group',
	onSelect: nop,
	onChange: nop,
	onKey: nop,
	value: EMPTY_STR,
	openOnClick: true,
	selectOnTab: false,
	suggestOn: 1,
	required: false,
	useKeys: true,
	arrow: true,
	close: true
};

export const SSuggestor = withClickOut(Suggestor);

export default SSuggestor;