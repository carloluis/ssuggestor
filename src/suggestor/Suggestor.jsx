import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { autoBind, keys, noop } from '../utils';
import transform from '../utils/transform';
import ListItem from './ListItem';

class Suggestor extends PureComponent {
	constructor(props) {
		super(props);
		autoBind(this);

		this.input = React.createRef();

		this.state = {
			filtered: this.filter(props.list, props.value, false),
			value: props.value,
			open: false,
			index: 0
		};
	}
	componentDidMount() {
		document.addEventListener('click', this._onClick);
	}
	componentWillUnmount() {
		document.removeEventListener('click', this._onClick);
	}
	_onClick(event) {
		if (!this.input.current.parentNode.contains(event.target)) {
			this.close();
		}
	}
	componentWillReceiveProps(nextProps) {
		let value = this.state.value;

		if (nextProps.value !== this.props.value && nextProps.value !== value) {
			value = nextProps.value;
		}

		this.setState({
			filtered: this.filter(nextProps.list, value, true),
			value
		});
	}
	close() {
		this.setState({
			open: false,
			filtered: this.unfilter(),
			index: 0
		});
	}
	handleClick() {
		if (this.props.openOnClick) {
			if (this.state.open) {
				this.close();
			} else {
				this.setState({ open: true, filtered: this.unfilter() });
			}
		}
	}
	handleKeyDown(e) {
		const { onKey, useKeys } = this.props;
		onKey(e);

		if (useKeys && this.processKey(e.keyCode)) {
			e.preventDefault();
		}
	}
	processKey(code) {
		const { open, index, filtered, value } = this.state;
		const ssuggestions = filtered.length ? filtered : this.unfilter();
		let nextIndex;

		switch (code) {
			case keys.ENTER:
				if (open && filtered[index]) {
					this.changeValue(filtered[index].word, true);
				} else {
					this.setState({ open: true, filtered: this.unfilter() });
				}
				break;
			case keys.ESCAPE:
				this.close();
				if (!open && value) {
					this.changeValue('');
				}
				break;
			case keys.DOWN:
				nextIndex = (index + open) % ssuggestions.length;
				break;
			case keys.UP:
				nextIndex = (index || ssuggestions.length) - 1;
				break;
			case keys.TAB:
				if (this.props.selectOnTab && open && filtered[index]) {
					this.changeValue(filtered[index].word, true);
				} else {
					this.close();
				}
			default:
				return false;
		}

		if (nextIndex !== undefined) {
			this.setState({ open: true, index: nextIndex, filtered: ssuggestions });
		}

		return true;
	}
	handleItemClick({ word }) {
		this.changeValue(word, true);
	}
	handleItemMouseEnter(index) {
		this.setState({ index });
	}
	handleChange(e) {
		e.stopPropagation();
		const value = e.target.value;
		this.changeValue(value);
	}
	remove() {
		this.changeValue('', true);
	}
	changeValue(value, select = false) {
		const { list, suggestOn, accents, onChange, onSelect } = this.props;
		const filtered = this.filter(list, value);
		const suggest = value.length >= suggestOn;
		const open = !!filtered.length && suggest;

		this.setState({ value, filtered, open }, () => {
			onChange(value);
			if (select) {
				const suggestion = filtered.find(({ word }) => transform(accents, word) === transform(accents, value));
				onSelect(value, suggestion && suggestion.item);
				this.close();
			}
		});
	}
	filter(list, value, onlyMatch = true) {
		const { accents, selector } = this.props;
		value = transform(accents, value);

		let mapped = list.map(item => {
			const word = selector(item);
			return {
				index: transform(accents, word).indexOf(value),
				word,
				item
			};
		});
		if (onlyMatch) {
			mapped = mapped.filter(item => item.index !== -1);
		}
		return mapped;
	}
	unfilter() {
		return this.filter(this.props.list, this.state.value, false);
	}
	focus() {
		this.input.current.focus();
	}
	render() {
		const { classSchema, style, placeholder, arrow, close, tooltip, required } = this.props;
		const { open, value, index, filtered } = this.state;
		const displaySuggestions = open && !!filtered.length;

		return (
			<div className={classSchema.root} onClick={this.handleClick} onKeyDown={this.handleKeyDown} style={style}>
				<input
					type="text"
					className={classSchema.input}
					onChange={this.handleChange}
					value={value}
					title={tooltip}
					placeholder={placeholder}
					required={required}
					ref={this.input}
				/>
				{arrow && <span className={classSchema.arrow} />}
				{close && value && <span className={classSchema.close} onClick={this.remove} />}
				{displaySuggestions && (
					<ul className={classSchema.list}>
						{filtered.map((item, i) => (
							<ListItem
								key={item.word}
								classSchema={classSchema}
								item={item}
								index={i}
								onItemClick={this.handleItemClick}
								onItemMouseEnter={this.handleItemMouseEnter}
								overItem={i === index}
								search={value}
							/>
						))}
					</ul>
				)}
			</div>
		);
	}
}

Suggestor.propTypes = {
	list: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.object])).isRequired,
	selector: PropTypes.func,
	onChange: PropTypes.func,
	onSelect: PropTypes.func,
	onKey: PropTypes.func,
	value: PropTypes.string,
	openOnClick: PropTypes.bool,
	selectOnTab: PropTypes.bool,
	placeholder: PropTypes.string,
	tooltip: PropTypes.string,
	classSchema: PropTypes.shape({
		root: PropTypes.string,
		arrow: PropTypes.string,
		close: PropTypes.string,
		list: PropTypes.string,
		item: PropTypes.string,
		activeItem: PropTypes.string
	}),
	suggestOn: PropTypes.number,
	style: PropTypes.object,
	required: PropTypes.bool,
	useKeys: PropTypes.bool,
	accents: PropTypes.bool,
	arrow: PropTypes.bool,
	close: PropTypes.bool
};

Suggestor.defaultProps = {
	classSchema: {},
	selector: s => s,
	onSelect: noop,
	onChange: noop,
	onKey: noop,
	value: '',
	openOnClick: true,
	selectOnTab: false,
	suggestOn: 1,
	required: false,
	accents: false,
	useKeys: true,
	arrow: true,
	close: true
};

export default Suggestor;
