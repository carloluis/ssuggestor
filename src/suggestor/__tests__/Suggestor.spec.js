import React from 'react';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';

import classSchema from '../../themes/bootstrap-3';
import * as utils from '../../utils';
import transform from '../../utils/transform';
import Suggestor from '../Suggestor';

const { keys, noop } = utils;

jest.mock('../../utils/transform', () => jest.fn((_, s) => s));
jest.mock('../../utils/noop', () => jest.fn());

const PROPS = {
	list: ['temporise', 'whencesoeve', 'turophile', 'umlaut'],
	onChange: jest.fn().mockName('oChangeMock'),
	onSelect: jest.fn().mockName('onSelectMock'),
	onKey: jest.fn(),
	value: '',
	openOnClick: false,
	selectOnTab: false,
	placeholder: 'placeholder text',
	tooltip: 'tooltip text',
	className: 'classname to apply',
	suggestOn: 1,
	style: { color: '#eee' },
	required: false,
	useKeys: false,
	accents: false,
	arrow: false,
	close: false,
	classSchema
};

describe('<Ssuggestor />', () => {
	it('snapshot - without suggestions', () => {
		const tree = renderer.create(<Suggestor {...PROPS} list={[]} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - with suggestions', () => {
		const tree = renderer.create(<Suggestor {...PROPS} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - with suggestions visible', () => {
		const tree = renderer.create(<Suggestor {...PROPS} openOnClick />);
		tree.getInstance().setState({ open: true });
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - with arrow', () => {
		const tree = renderer.create(<Suggestor {...PROPS} arrow />);
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - with close', () => {
		const tree = renderer.create(<Suggestor {...PROPS} close />);
		tree.getInstance().setState({ value: 'temp' });
		expect(tree).toMatchSnapshot();
	});
});

describe('Suggestor component', () => {
	beforeEach(() => {
		PROPS.onChange.mockReset();
		PROPS.onSelect.mockReset();
	});

	it('should call this.filter on component creation', () => {
		const filterSpy = jest.spyOn(Suggestor.prototype, 'filter');
		const component = shallow(<Suggestor {...PROPS} />);

		expect(filterSpy).toHaveBeenCalled();
	});

	describe('mounted component', () => {
		let mounted;
		beforeEach(() => {
			mounted = mount(<Suggestor {...PROPS} />);
		});

		it('should initialize state.open as false', () => {
			expect(mounted.state().open).toBeFalsy();
		});

		it('should filter list into initial state ', () => {
			expect(mounted.state()).toMatchObject({
				filtered: [
					{ word: 'temporise', index: 0 },
					{ word: 'whencesoeve', index: 0 },
					{ word: 'turophile', index: 0 },
					{ word: 'umlaut', index: 0 }
				],
				value: PROPS.value,
				index: 0
			});
		});

		it('should call changeValue when remove the value: remove -> changeValue', () => {
			// mounted.setProps({ openOnClick: true });
			const spy = jest.spyOn(mounted.instance(), 'changeValue');

			mounted.instance().remove();

			expect(spy).toBeCalled();
		});

		it('handleItemMouseEnter -> setState', () => {
			const spy = jest.spyOn(mounted.instance(), 'setState');

			mounted.instance().handleItemMouseEnter(2);

			expect(spy).toBeCalledWith({ index: 2 });
		});

		it('handleItemClick -> changeValue', () => {
			const spy = jest.spyOn(mounted.instance(), 'changeValue');
			const payload = { word: 'decrassify' };

			mounted.instance().handleItemClick(payload);

			expect(spy).toBeCalledWith(payload.word, true);
		});

		it('focus -> input.focus', () => {
			const spy = jest.spyOn(mounted.instance().input.current, 'focus');
			mounted.instance().focus();

			expect(spy).toBeCalled();
		});
	});

	describe('shallow component', () => {
		let component, instance;
		let setStateSpy, closeSpy, changeValueSpy;
		const autoBindSpy = jest.spyOn(utils, 'autoBind');

		beforeEach(() => {
			autoBindSpy.mockClear();
			component = shallow(<Suggestor {...PROPS} />);
			instance = component.instance();
			setStateSpy = jest.spyOn(instance, 'setState');
			closeSpy = jest.spyOn(instance, 'close');
			changeValueSpy = jest.spyOn(instance, 'changeValue');
		});

		afterEach(() => {
			expect(autoBindSpy).toHaveBeenCalledTimes(1);
		});

		it('changeValue -> [setState, props.onChange, props.onSelect]', () => {
			const value = 'umlaut';
			instance.changeValue(value, true);

			expect(setStateSpy).toBeCalled();
			expect(closeSpy).toBeCalledWith();
			expect(PROPS.onChange).toBeCalledWith(value);
			expect(PROPS.onSelect).toBeCalledWith(value, value);
		});

		it('handleChange -> changeValue', () => {
			const event = {
				stopPropagation: jest.fn(),
				target: {
					value: 'whencesoeve'
				}
			};
			instance.handleChange(event);

			expect(changeValueSpy).toBeCalled();
			expect(component.state()).toMatchObject({
				value: event.target.value,
				open: true
			});
		});
	});

	describe('componentWillReceiveProps', () => {
		let component, setStateSpy;

		beforeEach(() => {
			component = shallow(<Suggestor {...PROPS} />);
			setStateSpy = jest.spyOn(component.instance(), 'setState');

			const state = component.state();
			expect(state).toMatchObject({ value: '', filtered: expect.any(Array) });
			expect(state.filtered.length).toBe(PROPS.list.length);
		});

		afterEach(() => {
			expect(setStateSpy).toBeCalled();
		});

		it('update state.value if props.value changed', () => {
			const value = 'xesturgy';
			component.setProps({ value });

			expect(component.state()).toMatchObject({ value });
		});

		it('update state.filtered if props.list changed', () => {
			const list = ['one', 'two'];
			component.setProps({ list });

			expect(component.state().filtered.length).toBe(list.length);
		});

		it('should always call setState', () => {
			component.setProps({});
		});
	});

	describe('handleClick function', () => {
		let component, instance, closeSpy, setStateSpy;
		beforeEach(() => {
			component = shallow(<Suggestor {...PROPS} />);
			instance = component.instance();
			closeSpy = jest.spyOn(instance, 'close');
			setStateSpy = jest.spyOn(instance, 'setState');
		});

		it('should not perform any action (when openOnClick prop is falsy)', () => {
			instance.handleClick();
			expect(closeSpy).not.toBeCalled();
		});

		it('handleClick and openOnClick prop -> set open on when closed', () => {
			component.setProps({ openOnClick: true });
			instance.handleClick();

			expect(setStateSpy).toBeCalledWith({
				filtered: expect.any(Array),
				open: true
			});
			expect(closeSpy).not.toBeCalled();
		});

		it('handleClick and openOnClick prop -> toggle open flag', () => {
			component.setProps({ openOnClick: true });
			component.setState({ open: true });

			instance.handleClick();

			expect(setStateSpy).toBeCalledWith({
				filtered: expect.anything(),
				index: expect.any(Number),
				open: false
			});
			expect(closeSpy).toBeCalled();
		});
	});

	describe('... handleClose', () => {
		let mountedInstance, closeSpy;

		beforeEach(() => {
			mountedInstance = mount(<Suggestor {...PROPS} openOnClick />).instance();
			closeSpy = jest.spyOn(mountedInstance, 'close');
		});

		afterEach(() => {
			expect(closeSpy).toBeCalled();
		});

		it('should call when select value', () => {
			mountedInstance.changeValue('temp', true);
		});

		it('should call when _onClick', () => {
			mountedInstance.input = { current: { parentNode: { contains: jest.fn() } } };
			mountedInstance._onClick({ target: {} });
		});

		it('should call when open and click', () => {
			mountedInstance.handleClick();
			mountedInstance.handleClick();
		});
	});

	describe('... handles input changes', () => {
		const event = {
			stopPropagation: jest.fn(),
			target: { value: 'temp' }
		};
		let component;

		beforeEach(() => {
			event.stopPropagation.mockReset();
			PROPS.onChange.mockReset();
			component = shallow(<Suggestor {...PROPS} />);
		});

		it('should stop event propagation', () => {
			component.find('input').simulate('change', event);

			expect(event.stopPropagation).toBeCalled();
		});

		it('should update state', () => {
			component.find('input').simulate('change', event);

			expect(component.state().value).toBe(event.target.value);
			expect(component.state().open).toBeTruthy();
		});

		it('should call onChange prop callback', () => {
			component.find('input').simulate('change', event);

			expect(PROPS.onChange).toBeCalled();
			expect(PROPS.onSelect).not.toBeCalled();
		});

		it('should call onSelect prop callback when select value', () => {
			component.setProps({ openOnClick: true });
			component.instance().changeValue('temp', true);

			expect(PROPS.onChange).toBeCalled();
			expect(PROPS.onSelect).toBeCalled();
		});
	});

	describe('... handles keys', () => {
		const event = {
			preventDefault: jest.fn(),
			keyCode: keys.ENTER
		};
		let component;

		beforeEach(() => {
			event.preventDefault.mockReset();
			PROPS.onKey.mockReset();
			PROPS.onChange.mockReset();
			PROPS.onSelect.mockReset();

			component = shallow(<Suggestor {...PROPS} />);
		});

		it('should call prop onKey callback', () => {
			component.find('div').simulate('keyDown', event);

			expect(event.preventDefault).not.toBeCalled();
			expect(PROPS.onKey).toBeCalled();
		});

		it('should not call processKey if useKeys is falsy', () => {
			const processKeySpy = jest.spyOn(component.instance(), 'processKey');

			component.find('div').simulate('keyDown', { ...event, keyCode: keys.UP });

			expect(processKeySpy).not.toBeCalledWith();
		});

		describe('when using keys (useKeys: true)', () => {
			let processKeySpy, changeValueSpy, closeSpy, unfilterSpy, setStateSpy;

			beforeEach(() => {
				component.setProps({ useKeys: true });
				processKeySpy = jest.spyOn(component.instance(), 'processKey');
				changeValueSpy = jest.spyOn(component.instance(), 'changeValue');
				unfilterSpy = jest.spyOn(component.instance(), 'unfilter');
				closeSpy = jest.spyOn(component.instance(), 'close');
				setStateSpy = jest.spyOn(component.instance(), 'setState');
			});

			afterEach(() => {
				expect(PROPS.onKey).toBeCalled();
				expect(processKeySpy).toBeCalled();
			});

			it('should call processKey if useKeys is truthy', () => {
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.UP });
				expect(processKeySpy).toBeCalledWith(keys.UP);
			});

			it('should prevent event default action (with keys: ENTER, ESCAPE, UP, DOWN)', () => {
				expect(event.preventDefault).not.toBeCalled();
				const keyCodes = [keys.ENTER, keys.ESCAPE, keys.UP, keys.DOWN];

				keyCodes.map((keyCode, index) => {
					component.find('div').simulate('keyDown', { ...event, keyCode });
					expect(event.preventDefault).toHaveBeenCalledTimes(index + 1);
				});

				expect(event.preventDefault).toBeCalled();
				expect(event.preventDefault.mock.calls.length).toBe(4);
			});

			it('should not prevent event default action (with key: TAB)', () => {
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.TAB });
				expect(event.preventDefault).not.toBeCalled();
			});

			it('should only close suggestion list - if selectOnTab is off (TAB key)', () => {
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.TAB });

				expect(changeValueSpy).not.toBeCalled();
				expect(closeSpy).toBeCalled();
			});

			it('should change value (ENTER key)', () => {
				const selectedItem = PROPS.list[0];
				component.instance().setState({ open: true, index: 0 });

				component.find('div').simulate('keyDown', { ...event, keyCode: keys.ENTER });

				expect(event.preventDefault).toBeCalled();
				expect(changeValueSpy).toBeCalledWith(selectedItem, true);
				expect(PROPS.onSelect).toBeCalledWith(selectedItem, selectedItem);
				expect(unfilterSpy).toBeCalled();
			});

			it('should only call props.onKey and .processKey for unsupported key', () => {
				component.find('div').simulate('keyDown', { ...event, keyCode: 0 });

				expect(event.preventDefault).not.toBeCalled();
				expect(changeValueSpy).not.toBeCalled();
				expect(closeSpy).not.toBeCalled();
				expect(setStateSpy).not.toBeCalled();
				expect(unfilterSpy).not.toBeCalled();
				expect(PROPS.onSelect).not.toBeCalled();
			});

			it('should unfilter when no suggestions found for search and press [up]/[down] keys', () => {
				component.setProps({ openOnClick: true, useKeys: true });
				component.simulate('click');

				const stopPropagationMock = jest.fn();
				component.find('input').simulate('change', {
					stopPropagation: stopPropagationMock,
					target: { value: 'no-suggestion' }
				});

				component.simulate('keyDown', { ...event, keyCode: keys.DOWN });

				expect(stopPropagationMock).toBeCalled();
				expect(unfilterSpy).toBeCalled();
				expect(unfilterSpy.mock.calls.length).toBe(2);
			});

			it('should unfilter when no suggestions found for search and press [up] keys', () => {
				component.setProps({ openOnClick: true, useKeys: true });
				component.setState({ open: true });
				component.instance().changeValue('no-suggestions');
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.UP });

				expect(unfilterSpy).toBeCalled();
			});
		});

		describe('when keys and selectOnTab props are on', () => {
			let changeValueSpy, closeSpy;

			beforeEach(() => {
				component.setProps({ useKeys: true, selectOnTab: true });
				changeValueSpy = jest.spyOn(component.instance(), 'changeValue');
				closeSpy = jest.spyOn(component.instance(), 'close');
			});

			it('should change value (TAB key)', () => {
				component.setState({ open: true });
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.TAB });

				expect(changeValueSpy).toBeCalled();
				expect(closeSpy).toBeCalled();
				expect(PROPS.onChange).toBeCalled();
				expect(PROPS.onSelect).toBeCalled();
			});

			it('should clear selected value if any and suggestion list is closed (ESC key)', () => {
				component.setState({ open: false, value: 'temporise' });
				component.find('div').simulate('keyDown', { ...event, keyCode: keys.ESCAPE });

				expect(closeSpy).toBeCalled();
				expect(changeValueSpy).toBeCalledWith('');
				expect(PROPS.onChange).toBeCalled();
				expect(PROPS.onSelect).not.toBeCalled();
			});
		});
	});

	describe('filter', () => {
		let component, instance;

		beforeEach(() => {
			transform.mockClear();
			component = shallow(<Suggestor {...PROPS} />);
			instance = component.instance();
		});

		afterEach(() => {
			expect(transform).toBeCalled();
			transform.mockClear();
		});

		it('should call for every suggestion and the value', () => {
			expect(transform).toHaveBeenCalledTimes(PROPS.list.length + 1);
		});

		it('should call strip (if accents not allowed)', () => {
			instance.filter(PROPS.list, 'illaudable');

			expect(transform).toHaveBeenCalledTimes(2 * (PROPS.list.length + 1));
			transform.mock.calls.map(call => expect(call[0]).toBeFalsy());
		});

		it('should return all item on suggestion list (if onlyMatch arg set to falsy)', () => {
			const value = 'temporise';
			const result = instance.filter(PROPS.list, value, false);

			expect(result.length).toBe(4);

			const [first, ...tail] = result;

			expect(first).toMatchObject({
				word: value,
				index: 0
			});
			expect(tail.every(item => item.index === -1)).toBeTruthy();
		});

		it('should return only matches from suggestion list', () => {
			const result = instance.filter(PROPS.list, 'temporise');

			expect(result.length).toBe(1);
			expect(result[0].index).toBeGreaterThanOrEqual(0);
		});

		it('should return all if search pattern is empty', () => {
			const result = instance.filter(PROPS.list, '');

			expect(result.every(item => item.index === 0)).toBeTruthy();
			expect(result.length).toBe(4);
		});

		it('should not call strip (if accents support)', () => {
			component.setProps({ accents: true });
			transform.mockClear();

			component.instance().filter(PROPS.list, 'illaudable');

			expect(transform).toHaveBeenCalledTimes(PROPS.list.length + 1);
			transform.mock.calls.map(call => expect(call[0]).toBeTruthy());
		});
	});

	describe('unfilter', () => {
		const component = shallow(<Suggestor {...PROPS} />);

		it('should call component.filter', () => {
			const filterSpy = jest.spyOn(component.instance(), 'filter');
			component.instance().unfilter();

			expect(filterSpy).toHaveBeenCalledWith(PROPS.list, component.state().value, false);
		});
	});

	describe('cdm', () => {
		it('should add click listener to document', () => {
			const addEventListenerSpy = jest.spyOn(document, 'addEventListener');
			const component = shallow(<Suggestor {...PROPS} />);
			expect(addEventListenerSpy).toBeCalled();
		});
	});

	describe('cwu', () => {
		let component, cwuSpy;
		const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener');

		beforeEach(() => {
			removeEventListenerSpy.mockClear();
			component = shallow(<Suggestor {...PROPS} />);
			cwuSpy = jest.spyOn(component.instance(), 'componentWillUnmount');
			component.unmount();
		});

		it('should call componentWillUnmount', () => {
			expect(cwuSpy).toBeCalled();
		});

		it('should call removeEventListener', () => {
			expect(removeEventListenerSpy).toBeCalled();
		});

		it('should remove click listener from document', () => {
			expect(removeEventListenerSpy.mock.calls.length).toBe(1);
			expect(removeEventListenerSpy.mock.calls[0][0]).toBe('click');
		});
	});

	describe('_onClick', () => {
		const wrapper = mount(<Suggestor {...PROPS} />);
		const wrapperInstance = wrapper.instance();
		let closeSpy;

		beforeEach(() => {
			closeSpy = jest.spyOn(wrapperInstance, 'close');
			wrapper.update();
		});
		afterEach(() => {
			closeSpy.mockClear();
		});

		it('should not call handleClose when click inside component', () => {
			wrapperInstance._onClick({ target: wrapperInstance.input.current });
			expect(closeSpy).not.toBeCalled();
		});

		it('should call handleClose when click outside', () => {
			const contains = jest.fn(() => false);
			wrapperInstance.input = { current: { parentNode: { contains } } };
			wrapperInstance._onClick({ target: {} });

			expect(closeSpy).toBeCalled();
			expect(contains).toBeCalled();
		});
	});
});

describe('Suggestor - default cb props use noop', () => {
	let component, componentProps;

	beforeEach(() => {
		const { onChange, onSelect, onKey, ...props } = PROPS;
		component = mount(<Suggestor {...props} />);
		componentProps = component.props();
	});

	it('should set noop when not onChange, onSelect, oKey props func provided', () => {
		expect(componentProps.onChange).toBe(componentProps.onSelect);
		expect(componentProps.onSelect).toBe(componentProps.onKey);
	});

	it('sould call noop func', () => {
		expect(componentProps.onSelect()).toBe(undefined);
		expect(noop).toBeCalled();
	});
});

test('Suggestor renders suggestion list closed', () => {
	const component = ReactTestUtils.renderIntoDocument(<Suggestor {...PROPS} />);

	expect(component.state.open).toBeFalsy();
});
