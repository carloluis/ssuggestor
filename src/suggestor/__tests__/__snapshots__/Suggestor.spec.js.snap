// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ssuggestor /> snapshot - with arrow 1`] = `
<div
  className="input-group ss-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "color": "#eee",
    }
  }
>
  <input
    className="form-control ss-input"
    onChange={[Function]}
    placeholder="placeholder text"
    required={false}
    title="tooltip text"
    type="text"
    value=""
  />
  <span
    className="glyphicon glyphicon-triangle-bottom ss-triangle"
  />
</div>
`;

exports[`<Ssuggestor /> snapshot - with close 1`] = `
<div
  className="input-group ss-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "color": "#eee",
    }
  }
>
  <input
    className="form-control ss-input"
    onChange={[Function]}
    placeholder="placeholder text"
    required={false}
    title="tooltip text"
    type="text"
    value="temp"
  />
  <span
    className="glyphicon glyphicon-remove ss-remove"
    onClick={[Function]}
  />
</div>
`;

exports[`<Ssuggestor /> snapshot - with suggestions 1`] = `
<div
  className="input-group ss-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "color": "#eee",
    }
  }
>
  <input
    className="form-control ss-input"
    onChange={[Function]}
    placeholder="placeholder text"
    required={false}
    title="tooltip text"
    type="text"
    value=""
  />
</div>
`;

exports[`<Ssuggestor /> snapshot - with suggestions visible 1`] = `
<div
  className="input-group ss-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "color": "#eee",
    }
  }
>
  <input
    className="form-control ss-input"
    onChange={[Function]}
    placeholder="placeholder text"
    required={false}
    title="tooltip text"
    type="text"
    value=""
  />
  <ul
    className="dropdown-menu ss-list"
  >
    <li
      className="ss-over-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        temporise
      </a>
    </li>
    <li
      className=""
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        whencesoeve
      </a>
    </li>
    <li
      className=""
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        turophile
      </a>
    </li>
    <li
      className=""
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        umlaut
      </a>
    </li>
  </ul>
</div>
`;

exports[`<Ssuggestor /> snapshot - without suggestions 1`] = `
<div
  className="input-group ss-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "color": "#eee",
    }
  }
>
  <input
    className="form-control ss-input"
    onChange={[Function]}
    placeholder="placeholder text"
    required={false}
    title="tooltip text"
    type="text"
    value=""
  />
</div>
`;
